src/k8s-engine/__tests__/k3sHelper.spec.ts

```diff
@@ -265,4 +266,129 @@ describe(K3sHelper, () => {
+      });
+      test('picks same minor version after pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3', 'v1.5.0', 'v1.5.9']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+      test('picks same patch version before pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s8', 'v1.2.4+k3s1']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('picks same patch version after pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.4+k3s1', 'v1.2.4+k3s9']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+
+      test('picks same premajor version before pivot', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s2', 'v1.2.4+k3s1']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('can choose closest pre-release version for single digits', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s1', 'v1.2.3+k3s3', 'v1.2.3+k3s4']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+      test('verify single digits prereleases can pick the lower one', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s2', 'v1.2.3+k3s3', 'v1.2.3+k3s5']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[0]);
+      });
+      test('can choose closest pre-release version when crossing from single digits into 10s', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s3', 'v1.2.3+k3s8', 'v1.2.3+k3s12']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+    });
+
+    describe('selectClosestSemVer', () => {
+      // Note filenames are specified deliberately out of order in order to test sorting
+      test('finds the closest major version', () => {
+        const desiredSemver = new semver.SemVer('v3.1.2+k3s3');
+        const cachedFilenames = ['v1.2.9+k3s4', 'v4.2.8+k3s1', 'v4.3.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[1]);
+      });
+      test('finds the closest minor version', () => {
+        const desiredSemver = new semver.SemVer('v1.12.2+k3s3');
+        const cachedFilenames = ['v1.23.9+k3s4', 'v2.12.8+k3s1', 'v1.7.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[2]);
+      });
+      describe('finds the closest patch version', () => {
+        test('at the start of the list of existing names', () => {
+          const desiredSemver = new semver.SemVer('v1.12.2+k3s3');
+          const cachedFilenames = ['v1.12.9+k3s4', 'v1.12.8+k3s1', 'v1.12.4+k3s4', 'v1.12.4+k3s1'];
```

Comment:
We should include `v1.12.0` to check if it ends up picking the equidistant lower version.

Also, if I'm reading this correctly, the _start_ of the list is `v1.12.4+k3s1`, but we expect to pick `v1.12.4+k3s4` instead.

Reply:
I should actually have only one list unsorted -- it's easier to follow when the inputs are sorted.